compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -559,13 +563,118 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 let ty = self.tcx.erase_late_bound_regions(ty);
                 if self.can_coerce(expected, ty) {
                     err.span_label(sp, format!("expected `{}` because of return type", expected));
+                    self.try_suggest_return_impl_trait(err, expected, ty, fn_id);
                     return true;
                 }
                 false
             }
         }
     }
 
+    /// check whether the return type is a generic type with a trait bound
+    /// only suggest this if the generic param is not present in the arguments
+    /// if this is true, hint them towards changing the return type to `impl Trait`
+    /// ```
+    /// fn cant_name_it<T: Fn() -> u32>() -> T {
+    ///     || 3
+    /// }
+    /// ```
+    fn try_suggest_return_impl_trait(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+        fn_id: hir::HirId,
+    ) {
+        // Only apply the suggestion if:
+        //  - the return type is a generic parameter
+        //  - the generic param is not used as a fn param
+        //  - the generic param has at least one bound
+        //  - the generic param doesn't appear in any other bounds where it's not the Self type
+        // Suggest:
+        //  - Changing the return type to be `impl <all bounds>`
+
+        debug!("try_suggest_return_impl_trait, expected = {:?}, found = {:?}", expected, found);
+
+        let ty::Param(expected_ty_as_param) = expected.kind() else { return };
+
+        let fn_node = self.tcx.hir().find(fn_id);
+
+        let Some(hir::Node::Item(hir::Item {
+            kind:
+                hir::ItemKind::Fn(
+                    hir::FnSig { decl: hir::FnDecl { inputs: fn_parameters, output: fn_return, .. }, .. },
+                    hir::Generics { params, where_clause, .. },
+                    _body_id,
+                ),
+            ..
+        })) = fn_node else { return };
+
+        let Some(expected_generic_param) = params.get(expected_ty_as_param.index as usize) else { return };
+
+        // get all where BoundPredicates here, because they are used in to cases below
+        let where_predicates = where_clause.predicates.iter().filter_map(|p| match p {
+            WherePredicate::BoundPredicate(hir::WhereBoundPredicate {
+                bounds, bounded_ty, ..
+            }) => {
+                let ty = <dyn AstConv<'_>>::ast_ty_to_ty(self, bounded_ty);
+                Some((ty, bounds))
+            }
+            _ => None,
+        });
+
+        // check whether there is any predicate that contains our `T`, like `Option<T>: Send`
+        let any_predicate_contains_expected_ty =
+            where_predicates.clone().any(|(ty, _)| match ty.kind() {
+                ty::Param(param_ty) if param_ty == expected_ty_as_param => false,
+                _ => ty.contains(expected, self.tcx),
+            });
+
+        if any_predicate_contains_expected_ty {
```

Comment:
This doesn't necessarily cover cases where `T` appears in the bounds, right? So like `Option<()>: Trait<T>`

Reply:
Hmm, I wasn't able to figure out how I'd do that. Do you have an idea?

Reply:
Maybe instead of making `contains` be on Ty, you can make it a visitor. Then just visit both the ty and bounds.

Reply:
For the `contains` I was able to use the visitor on `Ty`. How should I visit the `PolyTraitRef` in the bounds? I found `hir::intravisit` but wasn't able to find a good way to use it.

Reply:
Sorry, thought I replied to this before.

There is a hir visitor: https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html

Alternatively, you *could* lower the trait ref.

I'm also fine it you just want to add this as a test case with a FIXME. This is just suggestion after all and doesn't have to he "perfect"

Reply:
I'll just leave a fixme in a test 👍